### PR TITLE
Add Discord Rich Presence support

### LINKED ISSUE #1109: Implement Discord Rich Presence into osu!
Rich Presence allows you to leverage the totally overhauled "Now Playing" section in a Discord user's profile to help people play your game together. Rich game data—including duration, score, current boss or map, and so much more—lives inside a user's profile alongside buttons that let you spectate or join a friend's game directly from Discord, or get a notification when it's over.

Rich Presence Image

_NOTE: The "Notify Me" feature is coming soon._

All you need is Discord's header file, and Discord's library.
In the header file, you'll find five event-emitting callbacks:

- ready()
- errored()
- disconnected()
- joinGame()
- spectateGame()

These five callbacks make up the entirety of what you need to implement. Behind the scenes, they do all the heavy lifting for you.

Learn more [here](https://discordapp.com/developers/docs/topics/rich-presence).

### PR DESCRIPTION
**Pull request by Game4all**
Closes #1109 and supersedes #4524

Prerequisites:
- [x] #4619 

Note that you'll have to change `client_id` in `DiscordRichPresenceClient` to the official osu! discord application client id and add the following to rich presence assets [rich_presence_icons.zip](https://github.com/ppy/osu/files/3101113/rich_presence_icons.zip)
with the following assets names: 

image

Video preview: https://streamable.com/p1m45

### CODE REVIEW

File: osu.Desktop/DiscordRichPresenceClient.cs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+namespace osu.Desktop
+{
+    internal class DiscordRichPresenceClient : Component
+    {
+        private const string client_id = "563024054391537674";
+
+        private Bindable<User> user;
+
+        private readonly DiscordRpcClient client = new DiscordRpcClient(client_id);
```
Comment (UselessToucan) [inline osu.Desktop/DiscordRichPresenceClient.cs]: `DiscordRpcClient` is `IDisposable`
see https://github.com/Lachee/discord-rpc-csharp/blob/master/DiscordRPC/DiscordRpcClient.cs

Hence I guess `DiscordRichPresenceClient` should implement `IDisposable` too.

File: osu.Desktop/DiscordRichPresenceClient.cs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        {
+            user = provider.LocalUser.GetBoundCopy();
+
+            user.ValueChanged += usr =>
+            {
+                usr.OldValue.Status.ValueChanged -= updateStatus;
+                usr.NewValue.Status.ValueChanged += updateStatus;
+            };
+
+            user.Value.Status.ValueChanged += updateStatus;
```
Comment (UselessToucan) [inline osu.Desktop/DiscordRichPresenceClient.cs]: Probably can be replaced with `user.Value.Status.TriggerValueChanged()`

File: osu.Desktop/DiscordRichPresenceClient.cs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+            client.OnReady += (_, __) => Logger.Log("Discord RPC Client ready.", LoggingTarget.Network, LogLevel.Debug);
```
Comment (UselessToucan) [inline osu.Desktop/DiscordRichPresenceClient.cs]: I suggest to extract a private method called `enableLogging` or something like this.

File: osu.Desktop/DiscordRichPresenceClient.cs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+                    presence.State = $"{game.Beatmap.Metadata.Artist} - {game.Beatmap.Metadata.Title} [{game.Beatmap.Version}]";
+                    setPresenceGamemode(game.Ruleset, presence);
+                    break;
+
+                case UserStatusEditing editing:
+                    presence.State = $"{editing.Beatmap.Metadata.Artist} - {editing.Beatmap.Metadata.Title} " + (!string.IsNullOrEmpty(editing.Beatmap.Version) ? $"[{editing.Beatmap.Version}]" : "");
+                    presence.Assets.SmallImageKey = "edit";
+                    presence.Assets.SmallImageText = "editing";
+                    break;
+            }
```
Comment (UselessToucan) [inline osu.Desktop/DiscordRichPresenceClient.cs]: Missing `default` block. We probably want to throw `ArgumentOutOfRangeException`.

**Review by UselessToucan — CHANGES_REQUESTED** [review]

**Comment by Game4all** [comment]
Due to changes on how user statuses are handled in #4619, logic for this needs to be 
rewritten.

File: osu.Desktop/DiscordRichPresenceClient.cs [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+            client.SetPresence(presence);
+        }
+
+        private void setPresenceGamemode(RulesetInfo ruleset, RichPresence presence)
+        {
+            switch (ruleset.ID)
+            {
+                case 0:
+                    presence.Assets.SmallImageKey = "osu";
```
Comment (peppy) [inline osu.Desktop/DiscordRichPresenceClient.cs]: `ruleset.ShortName`

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Desktop/DiscordRichPresenceClient.cs [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+                case 2:
+                    presence.Assets.SmallImageKey = ruleset.ShortName;
+                    break;
+
+                case 3:
+                    presence.Assets.SmallImageKey = ruleset.ShortName;
+                    break;
+
+                default:
```
Comment (hbnrmx) [inline osu.Desktop/DiscordRichPresenceClient.cs]: can be shortened to 
```
case 0:
case 1:
case 2:
case 3:
    //logic
```
Either way it's probably pointless to use a switch-case for just 2 different outcomes now

Comment (Game4all) [reply osu.Desktop/DiscordRichPresenceClient.cs]: You're right, I haven't thought of that. Let's use a if statement instead then

**Comment by peppy** [comment]
Please don't merge master unnecessarily.

**Comment by Game4all** [comment]
Sorry about that. I will run CI tests on my computer next time.

**Comment by LittleEndu** [comment]
Discord rpc is deprecated. The new gamesdk should be used instead. That being said, is this pr getting worked on or not?

**Comment by Game4all** [comment]
Updating to Discord GameSDK may be rather easy as it works the same way as old DiscordRPC library

However it could a bit more of work as the write of a C# FFI wrapper for the GameSDK is needed (adapting unity wrapper code?) as well as packaging native libraries for every desktop platform. Maybe worth to pack all this as a nuget package (to my knowledge nobody did this yet)

**Comment by LittleEndu** [comment]
I have a non-engine .NET app that uses the new gamesdk. All the C# code works fine out of the box, the only thing different from Unity are strings. Strings are UTF-16 in .NET while they are UTF-8 in Unity (discord devs are aware of this).

**Comment by Game4all** [comment]
Is it using a modified version of unity C# FFI ?

**Comment by LittleEndu** [comment]
Like I said, it's non-engine. It's using the latest .NET edge.js supports. I'm not quite sure what you are asking but I didn't need to do any modifications to what works in vanilla visual studio.

**Comment by bdach** [comment]
Call it an allergy if you will, but I'm *very* opposed to anything containing `js` in the name being integrated into the client.

**Comment by LittleEndu** [comment]
I used edge.js as a bridge from node.js to C# because discord gamesdk only supports C C# and C++. lazer being .NET to begin with wouldn't need to use any other libraries to implement gamesdk

**Comment by bdach** [comment]
Purely looking at the sdk code it does indeed have *some* chance of running without unity present, because unity-specific parts are hidden behind preprocessor directives and just won't compile (see *ImageManager* for an example). However it might be desired to have those actually work without unity, and it would just be saner to have the code packaged in a nuget instead of including files outright as @Game4all said.

**Comment by peppy** [comment]
Discord's official c# support is flakey as hell (especially with unicode string) so I can agree with the direction of this PR (Using the managed version of the library).

I'll see if we can get this in today as a baseline to build on.

**Review by peppy — DISMISSED** [review]

**Comment by peppy** [comment]
@smoogipoo requesting quick review as I changed >90% of code in original PR.

**Comment by LittleEndu** [comment]
#1109 shouldn't be closed until joinGame and spectateGame are also implemented. That or new issue should probably be opened for them.

**Comment by smoogipoo** [comment]
A new issue will have to be made for those parts of it.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**